SmartDeviceLink/SDLResponseDispatcher.m

```diff
@@ -86,67 +86,72 @@ - (instancetype)initWithNotificationDispatcher:(nullable id)dispatcher {
 #pragma mark - Storage
 
 - (void)storeRequest:(SDLRPCRequest *)request handler:(nullable SDLResponseHandler)handler {
-    NSNumber *correlationId = request.correlationID;
-
-    // Check for RPCs that require an extra handler
-    if ([request isKindOfClass:[SDLAddCommand class]]) {
-        SDLAddCommand *addCommand = (SDLAddCommand *)request;
-        if (!addCommand.cmdID) {
-            @throw [NSException sdl_missingIdException];
+    @synchronized (self) {
+        NSNumber *correlationId = request.correlationID;
+
+        // Check for RPCs that require an extra handler
+        if ([request isKindOfClass:[SDLAddCommand class]]) {
+            SDLAddCommand *addCommand = (SDLAddCommand *)request;
+            if (!addCommand.cmdID) {
+                @throw [NSException sdl_missingIdException];
+            }
+            if (addCommand.handler) {
+                self.commandHandlerMap[addCommand.cmdID] = addCommand.handler;
```

Comment:
The `sdl_runHandlerForCommand` method also accesses the `commandHandlerMap` but its access is not restricted.

Reply:
I re-did the whole sync model